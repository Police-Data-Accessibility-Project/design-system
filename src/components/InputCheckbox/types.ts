export interface PdapInputCheckboxProps {
	id: string;
	label?: string;
	name: string;
	defaultChecked?: boolean;
}
