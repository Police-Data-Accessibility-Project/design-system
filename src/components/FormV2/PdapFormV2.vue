<template>
	<form
		:id="id"
		:name="name"
		class="pdap-form"
		@submit.prevent="submit"
		@input.prevent="change"
		@change.prevent="change"
	>
		<slot v-if="$slots.error" name="error" />
		<div
			v-else-if="typeof errorMessage === 'string'"
			class="pdap-form-error-message"
		>
			{{ errorMessage }}
		</div>

		<slot />
	</form>
</template>
<script setup lang="ts">
// Globals
import { provide, ref, watchEffect } from 'vue';
import { useVuelidate } from '@vuelidate/core';

// Types
import { PdapFormPropsV2 } from './types';
import { provideKey, makeRules } from './util';

// Props
const props = defineProps<PdapFormPropsV2>();

// Emits
const emit = defineEmits(['submit', 'change', 'error']);

// Constants
const errorMessage = ref(props.error);
const values = ref(props.defaultValues ?? {});
const rules = props.schema ? makeRules(props.schema) : {};
const v$ = useVuelidate(rules, values, { $autoDirty: false, $lazy: true });

// Provide
provide(provideKey, {
	setValues,
	values,
	rules,
	v$,
});

// Expose
defineExpose({
	setValues,
	v$,
});

// Helpers
function setValues(val: Partial<typeof values.value>) {
	values.value = { ...values.value, ...val };
}
function resetForm() {
	v$.value.$reset();
	values.value = Object.entries(values.value).reduce((acc, [key]) => {
		return {
			...acc,
			[key]:
				typeof values.value[key] === 'string'
					? ''
					: Boolean(props.defaultValues?.[key]),
		};
	}, {});
}
function change(event: Event) {
	emit('change', values.value, event);
}

async function submit(e: Event) {
	// Check form submission
	const isValidSubmission = await v$.value.$validate();
	if (!isValidSubmission) emit('error', v$);
	if (isValidSubmission) {
		// Emit submit event (spread to new object to create new object, this allows us to reset `values` without messing with the data returned)
		emit('submit', { ...values.value }, e);

		resetForm();
	}
}

// Effects
// Effect - Updates form error state based on input error state and/or props
watchEffect(() => {
	if (props.error) errorMessage.value = props.error;
	else if (errorMessage.value && v$.value.$errors.length === 0)
		errorMessage.value = null;
	else if (!errorMessage.value && v$.value.$errors.length > 0)
		errorMessage.value = 'Please update this form to correct the errors';
});
</script>

<style>
@tailwind components;

@layer components {
	.pdap-form-error-message {
		@apply items-center justify-start basis-full flex-shrink flex bg-red-300 text-red-800 mb-3 p-2 text-sm;
	}
}
</style>
