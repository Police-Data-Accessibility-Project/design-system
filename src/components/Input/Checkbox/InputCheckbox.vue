<template>
	<input
		:id="id"
		class="pdap-input-checkbox"
		:name="name"
		:checked="value === 'true' || (!value && defaultChecked)"
		type="checkbox"
		:value="value"
		v-bind="$attrs"
		@change="change"
	/>
</template>

<script setup lang="ts">
import { PdapInputCheckboxProps } from '../types';

defineProps<PdapInputCheckboxProps>();

// Emits
const emit = defineEmits<{
	(event: 'change', val: Event): void;
}>();

const change = (event: Event) => {
	emit('change', event);
};
</script>
