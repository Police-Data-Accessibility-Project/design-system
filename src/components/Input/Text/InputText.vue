<template>
	<input
		:id="id"
		type="text"
		:name="name"
		:value="value"
		:placeholder="placeholder"
		@input="input"
	/>
</template>

<script setup lang="ts">
import { watchEffect } from 'vue';

// Types
import { type PdapInputTextProps } from '../types';

// Props
const props = withDefaults(defineProps<PdapInputTextProps>(), {});

// Emits
const emit = defineEmits<{
	(event: 'input', val: string): void;
}>();

const input = (event: Event) => {
	emit('input', (<HTMLInputElement>event.target).value);
};

watchEffect(() => {
	console.debug(
		`Hello from PDAP InputText ${props.name} rendered at ${window.location.pathname}`,
		{
			props,
		}
	);
});
</script>
