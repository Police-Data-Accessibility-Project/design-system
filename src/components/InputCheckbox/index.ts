export { default as InputCheckbox } from './PdapInputCheckbox.vue';
