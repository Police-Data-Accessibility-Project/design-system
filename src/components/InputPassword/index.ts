export { default as InputPassword } from './PdapInputPassword.vue';
