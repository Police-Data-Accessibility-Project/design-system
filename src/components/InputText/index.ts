export { default as InputText } from './PdapInputText.vue';
