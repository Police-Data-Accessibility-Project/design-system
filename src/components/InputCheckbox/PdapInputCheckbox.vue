<template>
	<div
		class="pdap-input pdap-input-checkbox"
		:class="{ ['pdap-input-error']: error }"
	>
		<slot v-if="$slots.error" name="error" class="pdap-input-error-message" />
		<div v-else-if="error" class="pdap-input-error-message">{{ error }}</div>

		<input
			:id="id"
			:name="name"
			:defaultChecked="defaultChecked"
			:value="values?.[name] ?? ''"
			v-bind="$attrs"
			type="checkbox"
			@input="onInput"
		/>

		<slot v-if="$slots.label" name="label" :for="id" />
		<label v-else-if="label" :for="id">{{ label }}</label>
	</div>
</template>

<script setup lang="ts">
import { computed, inject, useSlots } from 'vue';
import { PdapInputCheckboxProps } from './types';
import { PdapFormProvideV2 } from '../FormV2/types';
import { provideKey } from '../FormV2/util';

const { label, name } = defineProps<PdapInputCheckboxProps>();
const slots = useSlots();

if (!slots.label && !label)
	throw new Error(
		'All form inputs must have a label, passed as a slot or a prop'
	);

const { values, setValues, v$ } = inject<PdapFormProvideV2>(provideKey)!;

const error = computed(() => {
	return v$.value[name]?.$error ? v$.value[name]?.$errors?.[0]?.$message : '';
});

function onInput(e: Event) {
	setValues({ [name]: (e.target as unknown as HTMLInputElement).checked });
}
</script>
