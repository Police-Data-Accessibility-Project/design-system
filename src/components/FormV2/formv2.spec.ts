import { describe, it, expect, vi, beforeEach } from 'vitest';
import { nextTick } from 'vue';
import { mount } from '@vue/test-utils';
import PdapFormV2 from './PdapFormV2.vue';
import InputCheckbox from '../InputCheckbox/PdapInputCheckbox.vue';
import InputText from '../InputText/PdapInputText.vue';
import InputPassword from '../InputPassword/PdapInputPassword.vue';

vi.mock('vue-router');
vi.mock('vue', async () => {
	const actual: Record<string, unknown> = await vi.importActual('vue');
	return {
		...actual,
		/** Shim for `exportHelper` function which throws errors on Input because actual
		 * Vue impl does not have optional chaining on `sfc.__vccOpts` check */
		exportHelper: function (
			sfc: Record<string, unknown>,
			props: [string, string][]
		) {
			const newObject = sfc?.vccOpts ?? sfc;
			for (const [key, val] of props) {
				newObject[key] = val;
			}
			return newObject;
		},
	};
});
vi.mock('@vuelidate/core', async () => {
	const actual: Record<string, unknown> =
		await vi.importActual('@vuelidate/core');

	return {
		...actual,
	};
});

const submit = vi.fn((values: Record<PropertyKey, unknown>, e: Event) => ({
	values,
	e,
}));

const BASE_CONFIG = {
	props: {
		defaultValues: {
			name: '',
			email: '',
			password: '',
			'ice-cream': false,
		},
		schema: [
			{
				name: 'name',
				validators: {
					required: {
						value: true,
					},
				},
			},
			{
				name: 'email',
				validators: {
					required: { value: true },
					email: { value: true },
				},
			},
			{
				name: 'password',
				validators: {
					password: { value: true, message: 'Password is too weak' },
				},
			},
		],
		id: 'test',
		name: 'test',
	},
	attrs: {
		onSubmit: submit,
	},
	slots: {
		default: `
      <InputText
        id="name"
        name="name"
        label="Name"
        placeholder="Name"
      />
      <InputText
        id="email"
        name="email"
        label="Email"
        placeholder="Email"
      />
      <InputPassword
        id="password"
        name="password"
        label="Password"
        placeholder="Password"
      />
      <InputCheckbox
        id="ice-cream"
        name="ice-cream"
        label="Ice Cream"
        :defaultChecked="false"
      />
    `,
	},
	global: {
		stubs: {
			InputCheckbox,
			InputText,
			InputPassword,
		},
	},
};

describe('PdapFormV2', () => {
	let wrapper;

	beforeEach(() => {
		wrapper = mount(PdapFormV2, BASE_CONFIG);
	});

	it('renders the form element', () => {
		expect(wrapper.find('form').exists()).toBe(true);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('renders error message slot when provided', () => {
		wrapper = mount(PdapFormV2, {
			...BASE_CONFIG,
			slots: {
				error: '<div>Custom Error Message</div>',
			},
		});
		expect(wrapper.find('.pdap-form-error-message').exists()).toBe(false);
		expect(wrapper.text()).toContain('Custom Error Message');
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('renders error message when errorMessage prop is provided', () => {
		wrapper = mount(PdapFormV2, {
			...BASE_CONFIG,
			props: {
				...BASE_CONFIG.props,
				error: 'Form Error',
			},
		});
		expect(wrapper.find('.pdap-form-error-message').text()).toBe('Form Error');
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('renders default error message when form has errors', async () => {
		wrapper.find('form').trigger('submit');
		await nextTick();
		expect(wrapper.find('.pdap-form-error-message').text()).toBe(
			'Please update this form to correct the errors'
		);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('calls submit event with form values on valid submission', async () => {
		const form = await wrapper.find('form');
		await form.find('input[name="name"]').setValue('John Doe');
		await form.find('input[name="email"]').setValue('john@example.com');
		await form.find('input[name="password"]').setValue('Password123!');
		await form.find('input[name="ice-cream"]').setChecked();

		await form.trigger('submit');
		await wrapper.vm.$forceUpdate();
		await wrapper.vm.$nextTick();

		expect(submit).toHaveBeenCalledWith(
			{
				name: 'John Doe',
				email: 'john@example.com',
				password: 'Password123!',
				'ice-cream': true,
			},
			expect.any(Event)
		);
		expect(wrapper.html()).toMatchSnapshot();
	});
});
