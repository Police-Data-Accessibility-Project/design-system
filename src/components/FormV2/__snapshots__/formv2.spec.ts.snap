// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PdapFormV2 > calls submit event with form values on valid submission 1`] = `
<form class="pdap-form" id="test" name="test">
  <!--v-if-->
  <div class="pdap-input">
    <label for="name">Name</label>
    <!--v-if-->
    <input id="name" name="name" placeholder="Name" type="text">
  </div>
  <div class="pdap-input">
    <label for="email">Email</label>
    <!--v-if-->
    <input id="email" name="email" placeholder="Email" type="text">
  </div>
  <div class="pdap-input pdap-input-password">
    <label for="password">Password</label>
    <!--v-if-->
    <div class="pdap-input-password-wrapper">
      <input id="password" name="password" placeholder="Password" type="password">
      <button aria-label="Show text" class="pdap-input-password-toggle" type="button">
        <svg aria-hidden="true" class="svg-inline--fa fa-eye" data-icon="eye" data-prefix="fas" focusable="false" role="img" viewBox="0 0 576 512" xmlns="http://www.w3.org/2000/svg">
          <path class="" d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM144 256a144 144 0 1 1 288 0 144 144 0 1 1 -288 0zm144-64c0 35.3-28.7 64-64 64c-7.1 0-13.9-1.2-20.3-3.3c-5.5-1.8-11.9 1.6-11.7 7.4c.3 6.9 1.3 13.8 3.2 20.7c13.7 51.2 66.4 81.6 117.6 67.9s81.6-66.4 67.9-117.6c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3z" fill="currentColor" />
        </svg>
      </button>
    </div>
  </div>
  <div class="pdap-input pdap-input-checkbox">
    <!--v-if-->
    <input checked id="ice-cream" name="ice-cream" type="checkbox" value="true">
    <label for="ice-cream">Ice Cream</label>
  </div>
  <div class="pdap-input">
    <label for="notes">Notes?</label>
    <!--v-if-->
    <textarea id="notes" name="notes" placeholder="Notes" type="text"></textarea>
  </div>
</form>
`;

exports[`PdapFormV2 > renders default error message when form has errors 1`] = `
<form class="pdap-form" id="test" name="test">
  <div class="pdap-form-error-message">Please update this form to correct the errors</div>
  <div class="pdap-input pdap-input-error">
    <label for="name">Name</label>
    <div class="pdap-input-error-message">Value is required</div>
    <input id="name" name="name" placeholder="Name" type="text">
  </div>
  <div class="pdap-input pdap-input-error">
    <label for="email">Email</label>
    <div class="pdap-input-error-message">Value is required</div>
    <input id="email" name="email" placeholder="Email" type="text">
  </div>
  <div class="pdap-input pdap-input-password">
    <label for="password">Password</label>
    <!--v-if-->
    <div class="pdap-input-password-wrapper">
      <input id="password" name="password" placeholder="Password" type="password">
      <button aria-label="Show text" class="pdap-input-password-toggle" type="button">
        <svg aria-hidden="true" class="svg-inline--fa fa-eye" data-icon="eye" data-prefix="fas" focusable="false" role="img" viewBox="0 0 576 512" xmlns="http://www.w3.org/2000/svg">
          <path class="" d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM144 256a144 144 0 1 1 288 0 144 144 0 1 1 -288 0zm144-64c0 35.3-28.7 64-64 64c-7.1 0-13.9-1.2-20.3-3.3c-5.5-1.8-11.9 1.6-11.7 7.4c.3 6.9 1.3 13.8 3.2 20.7c13.7 51.2 66.4 81.6 117.6 67.9s81.6-66.4 67.9-117.6c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3z" fill="currentColor" />
        </svg>
      </button>
    </div>
  </div>
  <div class="pdap-input pdap-input-checkbox">
    <!--v-if-->
    <input id="ice-cream" name="ice-cream" type="checkbox" value="false">
    <label for="ice-cream">Ice Cream</label>
  </div>
  <div class="pdap-input pdap-input-error">
    <label for="notes">Notes?</label>
    <div class="pdap-input-error-message">Notes are required</div>
    <textarea id="notes" name="notes" placeholder="Notes" type="text"></textarea>
  </div>
</form>
`;

exports[`PdapFormV2 > renders error message slot when provided 1`] = `
<form class="pdap-form" id="test" name="test">
  <div>Custom Error Message</div>
</form>
`;

exports[`PdapFormV2 > renders error message when errorMessage prop is provided 1`] = `
<form class="pdap-form" id="test" name="test">
  <div class="pdap-form-error-message">Form Error</div>
  <div class="pdap-input">
    <label for="name">Name</label>
    <!--v-if-->
    <input id="name" name="name" placeholder="Name" type="text">
  </div>
  <div class="pdap-input">
    <label for="email">Email</label>
    <!--v-if-->
    <input id="email" name="email" placeholder="Email" type="text">
  </div>
  <div class="pdap-input pdap-input-password">
    <label for="password">Password</label>
    <!--v-if-->
    <div class="pdap-input-password-wrapper">
      <input id="password" name="password" placeholder="Password" type="password">
      <button aria-label="Show text" class="pdap-input-password-toggle" type="button">
        <svg aria-hidden="true" class="svg-inline--fa fa-eye" data-icon="eye" data-prefix="fas" focusable="false" role="img" viewBox="0 0 576 512" xmlns="http://www.w3.org/2000/svg">
          <path class="" d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM144 256a144 144 0 1 1 288 0 144 144 0 1 1 -288 0zm144-64c0 35.3-28.7 64-64 64c-7.1 0-13.9-1.2-20.3-3.3c-5.5-1.8-11.9 1.6-11.7 7.4c.3 6.9 1.3 13.8 3.2 20.7c13.7 51.2 66.4 81.6 117.6 67.9s81.6-66.4 67.9-117.6c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3z" fill="currentColor" />
        </svg>
      </button>
    </div>
  </div>
  <div class="pdap-input pdap-input-checkbox">
    <!--v-if-->
    <input id="ice-cream" name="ice-cream" type="checkbox" value="false">
    <label for="ice-cream">Ice Cream</label>
  </div>
  <div class="pdap-input">
    <label for="notes">Notes?</label>
    <!--v-if-->
    <textarea id="notes" name="notes" placeholder="Notes" type="text"></textarea>
  </div>
</form>
`;

exports[`PdapFormV2 > renders the form element 1`] = `
<form class="pdap-form" id="test" name="test">
  <!--v-if-->
  <div class="pdap-input">
    <label for="name">Name</label>
    <!--v-if-->
    <input id="name" name="name" placeholder="Name" type="text">
  </div>
  <div class="pdap-input">
    <label for="email">Email</label>
    <!--v-if-->
    <input id="email" name="email" placeholder="Email" type="text">
  </div>
  <div class="pdap-input pdap-input-password">
    <label for="password">Password</label>
    <!--v-if-->
    <div class="pdap-input-password-wrapper">
      <input id="password" name="password" placeholder="Password" type="password">
      <button aria-label="Show text" class="pdap-input-password-toggle" type="button">
        <svg aria-hidden="true" class="svg-inline--fa fa-eye" data-icon="eye" data-prefix="fas" focusable="false" role="img" viewBox="0 0 576 512" xmlns="http://www.w3.org/2000/svg">
          <path class="" d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM144 256a144 144 0 1 1 288 0 144 144 0 1 1 -288 0zm144-64c0 35.3-28.7 64-64 64c-7.1 0-13.9-1.2-20.3-3.3c-5.5-1.8-11.9 1.6-11.7 7.4c.3 6.9 1.3 13.8 3.2 20.7c13.7 51.2 66.4 81.6 117.6 67.9s81.6-66.4 67.9-117.6c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3z" fill="currentColor" />
        </svg>
      </button>
    </div>
  </div>
  <div class="pdap-input pdap-input-checkbox">
    <!--v-if-->
    <input id="ice-cream" name="ice-cream" type="checkbox" value="false">
    <label for="ice-cream">Ice Cream</label>
  </div>
  <div class="pdap-input">
    <label for="notes">Notes?</label>
    <!--v-if-->
    <textarea id="notes" name="notes" placeholder="Notes" type="text"></textarea>
  </div>
</form>
`;
